src/irmin-pack/s.ml

```diff
@@ -74,6 +74,62 @@ module type S = sig
 
   val stats :
     dump_blob_paths_to:string option -> commit:commit -> repo -> unit Lwt.t
+
+  module Snapshot : sig
+    type kinded_hash = Contents of hash * metadata | Node of hash
+    [@@deriving irmin]
+
+    type entry = { name : string; hash : kinded_hash } [@@deriving irmin]
+
+    type inode_tree = {
+      depth : int;
+      length : int;
+      pointers : (int * hash) list;
+    }
+    [@@deriving irmin]
+
+    type v = Inode_tree of inode_tree | Inode_value of entry list
+    [@@deriving irmin]
+
+    type inode = { v : v; root : bool } [@@deriving irmin]
+
+    type t = [ `Inode of inode | `Blob of Backend.Contents.Val.t ]
+    [@@deriving irmin]
+
+    type on_disk = [ `Reuse | `Path of string ]
+    (** The type for specifing an [on_disk] option when traversing a tree. The
+        traversal requires an index to keep track of visited elements.
+
+        - if [on_disk] is not specified, the index is in memory.
+        - if [on_disk] is [`Path path], the index is temporarily store at
+          [path].
+        - if [on_disk] is [`Reuse] the store's index is reused. *)
+
+    val export :
+      ?on_disk:on_disk ->
+      repo ->
+      (t -> unit Lwt.t) ->
+      root_key:Tree.kinded_key ->
+      int Lwt.t
+    (** [export ?on_disk repo f ~root_key] applies [f] to all inodes and
+        contents in a rooted tree, with root specified by [root_key]. All
+        elements are visited only once, in an unspecified but stable order.
+
+        Returns the total number of elements visited. *)
+
+    module Import : sig
+      type snapshot
+
+      val v : ?on_disk:on_disk -> repo -> snapshot
+      (** [v ?on_disk repo] create a [snaphot] instance. *)
+
+      val save_elt : snapshot -> t -> node_key Lwt.t
+      (** [save_elt snapshot elt] saves [elt] to the store. *)
+
+      val close : snapshot -> unit
+      (** [close snapshot] close the [snaphot] instance.*)
```

Comment:
Ideally, `v` would be the sole function here, and it would take a `t Seq.t`. Was this made impossible because of the integration in lib_context

Reply:
I tried using a seq here (it would have been nicer), but in `lib_context`, elements are read using `lwt_unix` ([here](https://gitlab.com/tezos/tezos/-/blob/master/src/lib_context/context_dump.ml#L167)) so instead of a seq we needed a sort of `lwt_seq`. I'm happy to try something else, if you have an idea